### PR TITLE
Modern gcc

### PR DESCRIPTION
**Pull request by PikalaxALT**
This serves as a parallel effort to [Ipatix's branch](/ipatix/feature/gcc7compat) and a revival of my own from September 2018, in which I

* Create a target to build Pokemon Emerald with CC1 from gcc-8.3.0 (as distributed with devkitARM-r52)
* Upgrade the source for compatibility with both agbcc and arm-none-eabi-gcc 8.3.0
* Upgrade gbafix to fix the ROM header in the ELF file (for no$gba-debug compat)

### CODE REVIEW

File: ld_script.txt [hunk 1]
```diff
@@ -515,12 +523,14 @@ SECTIONS {
         src/record_mixing.o(.rodata);
         src/secret_base.o(.rodata);
         src/tv.o(.rodata);
+		src/contest_link_80F57C4.o(.rodata);
```
Comment (ipatix) [inline ld_script.txt]: you might wanna fix indentation here

File: ld_script.txt [hunk 2]
```diff
@@ -1195,6 +1214,7 @@ SECTIONS {
     lib_rodata :
     SUBALIGN(4)
     {
+		src/m4a.o(.rodata);
```
Comment (ipatix) [inline ld_script.txt]: same here

File: src/main.c [hunk 3]
```diff
@@ -85,7 +85,7 @@ void EnableVCountIntrAtLine150(void);
 
 void AgbMain()
 {
-    RegisterRamReset(RESET_ALL);
+    RegisterRamReset(MODERN ? RESET_ALL & ~RESET_IWRAM : RESET_ALL);
```
Comment (ipatix) [inline src/main.c]: You might want to comment this switch. Could be pretty confusing if you don't know why you have to do it.

Comment (nullableVoidPtr) [reply src/main.c]: `//Exclude IWRAM from resetting if compiled with modern gcc`?

Comment (PikalaxALT) [reply src/main.c]: Observe the "outdated" tag please

Comment (ipatix) [reply src/main.c]: Oh yeah, sorry, didn't know that tag was a thing. Thanks

File: tools/gbafix/gbafix.c [hunk 4]
```diff
@@ -138,6 +138,7 @@ int main(int argc, char *argv[])
     char *argfile = 0;
     FILE *infile;
     int silent = 0;
+    int schedule_pad = 0;
```
Comment (huderlem) [inline tools/gbafix/gbafix.c]: I would suggest upstreaming this to devkitpro.  I believe the only reason this copy exists is because there was a bugfix was needed at the time, and we didn't want to wait for another dkp release.

File: INSTALL.md [hunk 5]
```diff
@@ -101,3 +101,16 @@ If you've only changed `.c` or `.s` files, you can turn off the dependency scann
...
+
+This project supports the `arm-none-eabi-gcc` compiler which ships with devkitARM r52.  To build this target, simply run:
+
+    make modern
+
+# Building with your own toolchain
+
+To build Pokemon Emerald with a toolchain other than devkitARM, override the `DEVKITARM` environment variable with the path to your toolchain.  Example:
+
+    make compare DEVKITARM=/usr/local/arm-none-eabi
```
Comment (huderlem) [inline INSTALL.md]: Overriding `DEVKITARM` doesn't sit well with me.  When you override it, it's no longer pointing to anything `DEVKITARM`-related, so I guess I would suggest defining a different environment variable.  I think it has potential to cause lots of confusion (in the Makefile, and users of the repo)

File: Makefile [hunk 6]
```diff
@@ -109,7 +132,10 @@ clean: tidy
 
 tidy:
 	rm -f $(ROM) $(ELF) $(MAP)
+ifeq ($(MODERN),0)
+	@$(MAKE) tidy MODERN=1
 	rm -r build/*
```
Comment (huderlem) [inline Makefile]: I think this should be `rm -r $(OBJ_DIR)`?

File: tools/gbafix/gbafix.c (hunk 4 above)
Comment (PikalaxALT) [reply tools/gbafix/gbafix.c]: Sure, if they'll take our advice.  Should I bump the version?

Comment (huderlem) [reply tools/gbafix/gbafix.c]: They seem to bump the version every commit, so I would guess yes.

**Comment by ipatix** [comment]
@PikalaxALT  Does your current version also work without -mabi=gnu-apcs?

File: tools/gbafix/gbafix.c (hunk 4 above)
Comment (PikalaxALT) [reply tools/gbafix/gbafix.c]: Bumped the version number and updated the changelog.

File: Makefile (hunk 6 above)
Comment (PikalaxALT) [reply Makefile]: Fixed

**Comment by PikalaxALT** [comment]
@ipatix 
```
  Known ARM ABIs (for use with the -mabi= option):
    aapcs aapcs-linux apcs-gnu atpcs iwmmxt
```

`gnu-apcs` is not a valid value.

**Comment by ipatix** [comment]
Yeah, my bad, I was referring to apcs-gnu.

File: src/pokemon_animation.c [hunk 7]
```diff
@@ -861,16 +861,22 @@ u8 GetSpeciesBackAnimSet(u16 species)
...
+#define tPtrHi  data[1]
+#define tPtrLo  data[2]
 #define tAnimId data[3]
 #define tSaved0 data[4]
 #define tSaved2 data[5]
 
+#if MODERN
+#define ANIM_SPRITE(taskId)   ((struct Sprite *)((gTasks[taskId].tPtrHi << 16) | ((u16)gTasks[taskId].tPtrLo)))
+#else
+#define ANIM_SPRITE(taskId)   ((struct Sprite *)((gTasks[taskId].tPtrHi << 16) | (gTasks[taskId].tPtrLo)))
```
Comment (SBird1337) [inline src/pokemon_animation.c]: Could you add a comment that explains that this is buggy under some circumstances?

Comment (PikalaxALT) [reply src/pokemon_animation.c]: Resolved.

File: src/party_menu.c [hunk 8]
```diff
@@ -1024,12 +1024,66 @@ static const u8 *const sSelectionStringTable[] =
 };
 
 // Unknown unused data. Feel free to remove.
-static const u16 gUnknown_08615B94[] =
```
Comment (ipatix) [inline src/party_menu.c]: What does this have to do with modern gcc?

Comment (PikalaxALT) [reply src/party_menu.c]: whoops

also, i have no idea why it's FAILED; it's OK on my end :thonk:

**Comment by PikalaxALT** [comment]
"Not OK" status cannot be replicated locally.

**Comment by PikalaxALT** [comment]
Please ignore all Travis failure notifications until further notice, as there is an unknown degradation on Travis.
